README.md

```diff
@@ -803,15 +823,47 @@ CREATE TRIGGER generate_pdf_update
   EXECUTE PROCEDURE trigger_job('generate_pdf');
 ```
 
-## Updating and removing jobs
+## Replacing, updating and removing jobs
 
-Jobs scheduled with a `job_key` parameter may be updated later, provided they
+### Replacing/updating jobs
+
+Jobs scheduled with a `job_key` parameter may be replaced/updated, provided they
 are still pending, by calling `add_job` again with the same `job_key` value.
-This can be used for rescheduling jobs or to ensure only one of a given job is
-scheduled at a time. When a job is updated, any omitted parameters are reset to
-their defaults, with the exception of `queue_name` which persists unless
-overridden. For example after the below SQL transaction, the `send_email` job
-will run only once, with the payload `'{"count": 2}'`:
+This can be used for rescheduling jobs, to ensure only one of a given job is
+scheduled at a time, or to update other settings for the job. Behavior of this
+updating is controlled by the `job_key_mode` setting, but in all cases the
+following is true:
+
+- If an existing job with the same job key is found:
+  - If the existing job is locked:
+    - it will have it's `key` cleared
+    - it will have its attempts set to `max_attempts` to avoid it running again
+    - a new job will be created with the given attributes.
+  - Otherwise, if the existing job has previously failed:
```

Comment:
As a clue: this approach imo potentially complicates tracking failed jobs in a pipe as job as an entity, identified by some unique id can suddenly transfer from state with non-zero amount of fails to straight successful and confuse metrics (specifically calculating failed jobs and job errors). I might be wrong due to vague arch understanding.

Reply:
Good feedback; the reason I've gone with this approach is that "failed forever" tasks are expensive (they sit around forever making every scan for new jobs more expensive as the table is larger): they should be avoided where possible and periodically purged where not possible. This approach has been taken to improve the performance of the system, it's advised that jobs include their own logic for error tracking (or use the new events system #155) rather than scanning the database itself which will impact performance. Reporting systems can track the full job if they want (through the event system) and can see the attempts "drop down" to zero and track as appropriate.